server/src/main/java/org/elasticsearch/cluster/metadata/Metadata.java

```diff
@@ -1349,10 +1349,30 @@ public Metadata build() {
 
         private SortedMap<String, IndexAbstraction> buildIndicesLookup() {
             SortedMap<String, IndexAbstraction> indicesLookup = new TreeMap<>();
+            Map<String, DataStream> indexToDataStreamLookup = new HashMap<>();
+            DataStreamMetadata dataStreamMetadata = (DataStreamMetadata) this.customs.get(DataStreamMetadata.TYPE);
+            if (dataStreamMetadata != null) {
+                for (DataStream ds : dataStreamMetadata.dataStreams().values()) {
+                    for (Index i : ds.getIndices()) {
+                        indexToDataStreamLookup.put(i.getName(), ds);
+                    }
+                }
+            }
+
             for (ObjectCursor<IndexMetadata> cursor : indices.values()) {
                 IndexMetadata indexMetadata = cursor.value;
-                IndexAbstraction existing =
-                    indicesLookup.put(indexMetadata.getIndex().getName(), new IndexAbstraction.Index(indexMetadata));
+
+                IndexAbstraction.Index index;
+                DataStream parent = indexToDataStreamLookup.get(indexMetadata.getIndex().getName());
+                if (parent != null) {
+                    List<IndexMetadata> backingIndices = parent.getIndices().stream()
```

Comment:
Can we add:

```
assert parent.getIndices().contains(indexMetadata.getIndex())
```

that would  validate that uuid and name from data-stream indices is the same as the index we are looking at.